draft-ietf-quic-load-balancers.md

```diff
@@ -436,91 +436,158 @@ for decryption. To understand this algorithm, it is useful to define four
 functions that minimize the amount of bit-shifting necessary in the event that
 there are an odd number of octets.
 
-The expand_left() function outputs 16 octets, with its first argument in the
-most significant bits, its second argument in the least significant byte, its
-third argument in the second least significant byte, and zeros in all other
-positions. Thus,
+When configured with both a key, and a nonce length and server ID length that
+sum to any number other than 16, the server MUST follow the algorith below to
+encrypt the connection ID.
 
-~~~pseudocode
-expand_left(0xaaba3c, 0x0b, 0x02) =
-                 0xaaba3c0000000000000000000000020b
+#### Overview
+
+The 4-pass algorithm is a four-round Feistel Network with the round function
+being AES-ECB. Most modern applications of Feistel Networks have more than four
+rounds. The implications of this choice, which is meant to limit the per-packet
+compute overhead at load balancers, are discussed in
+{{distinguishing-attacks}}.
+
+The server concatenates the server ID and nonce into a single field, which is
+then split into equal halves. In successive passes, one of these halves is
+expanded into a 16B plaintext, encrypted with AES-ECB, and the result XORed with
+the other half. The diagram below shows the conceptual processing of a plaintext
+server ID and nonce into a connection ID. 'FO' stands for 'First Octet'.
+
+~~~
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   | FO  | Server ID |         Nonce         |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |                    |
+      |                    \/
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |  |      left_0     |      right_0    |
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |    |    |                      |
+      |    |    \/                     |
+      |    |  +-+-+-+-+-+              \/
+      |    |  | AES-ECB |------------>(+)
+      |    |  +-+-+-+-+-+              |
+      |    \/         +-+-+-+-+-+      | right_1
+      |   (+)<--------| AES-ECB |<-----|
+      |    |          +-+-+-+-+-+      |
+      |    | left_1   +-+-+-+-+-+      \/
+      |    |--------->| AES-ECB |---->(+)
+      |    |          +-+-+-+-+-+      |
+      |    \/         +-+-+-+-+-+      | right_2
+      |   (+)<--------| AES-ECB |------|
+      |    |          +-+-+-+-+-+      |
+      |    \/                          \/
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |  |      left_2     |      right_2    |
+      |  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |          |                  |
+      \/         \/                 \/
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   | FO  |            Ciphertext             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

Comment:
```suggestion
~~~ aasvg
   +-----+-----------+-----------------------+
   | FO  | Server ID |         Nonce         |
   +--+--+-----------+-----+-----------------+
      |                    |
      |                    V
      |  +-----------------+-----------------+
      |  |      left_0     |      right_0    |
      |  +--+--------------+--------------+--+
      |     |                             |
      |     |                             |
      |     |         .--------.          V
      |     +-------->| AES-ECB +-------->⊕
      |     |         '--------'          |
      |     V             .--------.      | right_1
      |     ⊕<-----------+ AES-ECB |<-----+
      |     |             '--------'      |
      |     | left_1  .--------.          V
      |     +-------->| AES-ECB +-------->⊕
      |     |         '--------'          |
      |     V             .--------.      |
      |     ⊕<-----------+ AES-ECB |<-----+
      |     |             '--------'      |
      |     |                             |
      |     V                             V
      |  +-----------------+-----------------+
      |  |      left_2     |      right_2    |
      |  +-------+---------+--------+--------+
      |          |                  |
      V          V                  V
   +-----+-----------------------------------+
   | FO  |            Ciphertext             |
   +-----+-----------------------------------+
```

This looks something like this:

![image](https://user-images.githubusercontent.com/67641/197428830-b24c676a-ae63-428c-a300-d46defa4b4cf.png)